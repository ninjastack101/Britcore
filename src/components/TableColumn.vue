<template>
  <!-- Never render the contents -->
  <!-- The scoped slot won't have the required data -->
  <div v-if="false">
    <slot></slot>
  </div>
</template>

<script>
export default {
  name: "table-column",

  props: {
    show: { required: false, type: String },
    label: { default: null, type: String },
    dataType: { default: "string", type: String },

    sortable: { default: true, type: Boolean },
    sortBy: { default: null },

    filterable: { default: true, type: Boolean },
    filterOn: { default: null },

    formatter: { default: v => v, type: Function },

    hidden: { default: false, type: Boolean },

    cellClass: { default: "" },
    headerClass: { default: "" }
  }
};
</script>
