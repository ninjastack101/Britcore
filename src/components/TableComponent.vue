<template>
  <div class="table-component">
    <div v-if="showFilter && filterableColumnExists" class="table-component__filter">
      <input
        :class="fullFilterInputClass"
        type="text"
        v-model="filter"
        :placeholder="filterPlaceholder"
      >
      <a v-if="filter" @click="filter = ''" class="table-component__filter__clear">×</a>
    </div>

    <div class="table-component__table-wrapper">
      <table :class="fullTableClass">
        <caption
          v-if="showCaption"
          class="table-component__table__caption"
          role="alert"
          aria-live="polite"
        >{{ ariaCaption }}</caption>
        <thead :class="fullTableHeadClass">
          <tr>
            <table-column-header
              @click="changeSorting"
              v-for="column in columns"
              :key="column.show"
              :sort="sort"
              :column="column"
            ></table-column-header>
          </tr>
        </thead>
        <tbody :class="fullTableBodyClass">
          <table-row
            v-for="row in displayedRows"
            :key="row.vueTableComponentInternalRowId"
            :row="row"
            :columns="columns"
            @rowClick="emitRowClick"
          ></table-row>
        </tbody>
        <tfoot>
          <slot name="tfoot" :rows="rows"></slot>
        </tfoot>
      </table>
    </div>

    <div v-if="displayedRows.length === 0" class="table-component__message">{{ filterNoResults }}</div>

    <div style="display:none">
      <slot></slot>
    </div>
  </div>
</template>

<script>
import Column from "../classes/Column";
import Row from "../classes/Row";
import TableColumnHeader from "./TableColumnHeader";
import TableRow from "./TableRow";
import { classList } from "../helpers";

export default {
  name: "table-component",
  components: {
    TableColumnHeader,
    TableRow
  },

  props: {
    data: { default: () => [], type: [Array, Function] },

    showFilter: { default: true },
    showCaption: { default: true },

    sortBy: { default: "", type: String },
    sortOrder: { default: "", type: String },

    tableClass: { default: () => "" },
    theadClass: { default: () => "" },
    tbodyClass: { default: () => "" },
    filterInputClass: { default: () => "" },
    filterPlaceholder: { default: () => "" },
    filterNoResults: { default: () => "" }
  },

  data: () => ({
    columns: [],
    rows: [],
    filter: "",
    sort: {
      fieldName: "",
      order: ""
    },
    localSettings: {}
  }),

  created() {
    this.sort.fieldName = this.sortBy;
    this.sort.order = this.sortOrder;
  },

  async mounted() {
    const columnComponents = this.$slots.default
      .filter(column => column.componentInstance)
      .map(column => column.componentInstance);

    this.columns = columnComponents.map(column => new Column(column));

    columnComponents.forEach(columnCom => {
      Object.keys(columnCom.$options.props).forEach(prop =>
        columnCom.$watch(prop, () => {
          this.columns = columnComponents.map(column => new Column(column));
        })
      );
    });
    await this.mapDataToRows();
  },

  watch: {
    filter() {
      if (!this.usesLocalData) {
        this.mapDataToRows();
      }
    },

    data() {
      if (this.usesLocalData) {
        this.mapDataToRows();
      }
    }
  },

  computed: {
    fullTableClass() {
      return classList("table-component__table", this.tableClass);
    },

    fullTableHeadClass() {
      return classList("table-component__table__head", this.theadClass);
    },

    fullTableBodyClass() {
      return classList("table-component__table__body", this.tbodyClass);
    },

    fullFilterInputClass() {
      return classList("table-component__filter__field", this.filterInputClass);
    },

    ariaCaption() {
      if (this.sort.fieldName === "") {
        return "Table not sorted";
      }

      return (
        `Table sorted by ${this.sort.fieldName} ` +
        (this.sort.order === "asc" ? "(ascending)" : "(descending)")
      );
    },

    usesLocalData() {
      return Array.isArray(this.data);
    },

    displayedRows() {
      if (!this.usesLocalData) {
        return this.sortedRows;
      }

      if (!this.showFilter) {
        return this.sortedRows;
      }

      if (!this.columns.filter(column => column.isFilterable()).length) {
        return this.sortedRows;
      }

      return this.sortedRows.filter(row => row.passesFilter(this.filter));
    },

    sortedRows() {
      if (!this.usesLocalData) {
        return this.rows;
      }

      if (this.sort.fieldName === "") {
        return this.rows;
      }

      if (this.columns.length === 0) {
        return this.rows;
      }

      const sortColumn = this.getColumn(this.sort.fieldName);

      if (!sortColumn) {
        return this.rows;
      }
      // eslint-disable-next-line
      return this.rows.sort(
        sortColumn.getSortPredicate(this.sort.order, this.columns)
      );
    },

    filterableColumnExists() {
      return this.columns.filter(c => c.isFilterable()).length > 0;
    }
  },

  methods: {
    async mapDataToRows() {
      const data = this.prepareLocalData();

      let rowId = 0;

      this.rows = data
        .map(rowData => {
          rowData.vueTableComponentInternalRowId = rowId++;
          return rowData;
        })
        .map(rowData => new Row(rowData, this.columns));
    },

    prepareLocalData() {
      return this.data;
    },

    changeSorting(column) {
      if (this.sort.fieldName !== column.show) {
        this.sort.fieldName = column.show;
        this.sort.order = "asc";
      } else {
        this.sort.order = this.sort.order === "asc" ? "desc" : "asc";
      }
    },

    getColumn(columnName) {
      return this.columns.find(column => column.show === columnName);
    },

    emitRowClick(row) {
      this.$emit("rowClick", row);
      this.$emit("row-click", row);
    }
  }
};
</script>
<style>
.table-component__table td {
  padding: 0.75em 1.25em;
  vertical-align: top;
  text-align: left;
  padding: 15px;
}
</style>

<style scoped>
*,
*:after,
*:before {
  position: relative;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.table-component {
  display: flex;
  flex-direction: column;
  margin: 4em 0;
}

.table-component__filter {
  align-self: flex-end;
}

.table-component__filter__field {
  padding: 0 1.25em 0 0.75em;
  height: 2.5em;
  border: solid 2px #e0e0e0;
  border-radius: 2em;
  font-size: inherit;
}

.table-component__filter__clear {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 2em;
  color: #007593;
  font-weight: bold;
  cursor: pointer;
}

.table-component__filter__field:focus {
  outline: 0;
  border-color: #007593;
}

.table-component__table-wrapper {
  overflow-x: auto;
  margin: 1em 0;
  width: 100%;
  border: solid 1px #ddd;
  border-bottom: none;
}

.table-component__table {
  min-width: 100%;
  border-collapse: collapse;
  border-bottom: solid 1px #ddd;
  table-layout: fixed;
}

.table-component__table__caption {
  position: absolute;
  top: auto;
  left: -10000px;
  overflow: hidden;
  width: 1px;
  height: 1px;
}

.table-component__table th,
.table-component__table td {
  padding: 0.75em 1.25em;
  vertical-align: top;
  text-align: left;
  padding: 15px;
}

.table-component__table th {
  background-color: #e0e0e0;
  color: #999;
  text-transform: uppercase;
  white-space: nowrap;
  background: #f8f9fa;
  color: #01798e;
  border-bottom: 2px solid #dee2e6;
}

.table-component__table tbody tr:nth-child(even) {
  background-color: #f0f0f0;
}

.table-component__table a {
  color: #007593;
}

.table-component__message {
  color: #999;
  font-style: italic;
}

.table-component__th--sort,
.table-component__th--sort-asc,
.table-component__th--sort-desc {
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.table-component__th--sort-asc:after,
.table-component__th--sort-desc:after {
  position: absolute;
  left: 0.25em;
  display: inline-block;
  color: #bbb;
}

.table-component__th--sort-asc:after {
  content: "↑";
}

.table-component__th--sort-desc:after {
  content: "↓";
}
</style>
