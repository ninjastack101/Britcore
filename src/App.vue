<template>
  <div id="appTable">
    <table-component :data="getData()" sort-by="name" sort-order="asc" ref="table" filter-placeholder="Filter data...">
      <table-column :sortable="false" show="id" label="ID"></table-column>
      <table-column show="name" label="Name"></table-column>
      <table-column show="description" label="Description"></table-column>
      <table-column show="date" label="Date" data-type="date:YYYY-MM-DDTHH:mm:ss"></table-column>
      <table-column show="amount" label="Amount" data-type="numeric"></table-column>
      <table-column :sortable="false" :filterable="false">
        <template slot-scope="payment">
          <button v-on:click="openModal(payment)">Edit</button>
        </template>
      </table-column>
    </table-component>
    <edit-row
      v-if="showModal"
      :payment="currentSelectedPayment"
      @close="showModal = false"
      @submit="updatePayment"
    ></edit-row>
  </div>
</template>

<script>
import TableComponent from "./components/TableComponent";
import TableColumn from "./components/TableColumn";
import EditRow from "./components/EditRow";

export default {
  name: "App",
  components: {
    TableComponent,
    TableColumn,
    EditRow
  },

  data: () => ({
    showModal: false,
    currentSelectedPayment: {}
  }),

  methods: {
    getData() {
      return this.$store.state.users;
    },

    openModal(payment) {
      this.showModal = true;
      this.currentSelectedPayment = payment;
    },

    updatePayment(payment) {
      this.$store.dispatch("updateUser", payment);
      this.showModal = false;
    }
  },
  mounted() {
    this.$store.dispatch("getUsers");
  }
};
</script>
