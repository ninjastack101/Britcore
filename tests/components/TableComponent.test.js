import Vue from 'vue/dist/vue.js';
import simulant from 'simulant';
import App from '../../src/App';
import TableComponent from '../../src/components/TableComponent';
import TableColumn from '../../src/components/TableColumn';

describe('TableComponent', () => {

  it('can mount', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John' },{ firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('can display nested properties', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{nested: { firstName: 'John' } }, { nested: { firstName: 'Paul' }}]">
                    <table-column show="nested.firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });


  it('supports a scoped slot inside the table column', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John' },{ firstName: 'Paul' }]">
                    <table-column label="First name">
                        <template slot-scope="row">
                           {{ row.firstName }} slot
                        </template>
                    </table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('supports a named slot to display a tfoot section', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John' },{ firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                    <template slot="tfoot" slot-scope="{ rows }">
                        <tr>
                            <td>Name count:</td>
                            <td>{{ rows.length }}</td>
                        </tr>
                    </template>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('has an prop to disable the filter', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component :show-filter="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('has an prop to disable the caption', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component :show-caption="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('will use the property name as a column heading if label is not set', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component :show-filter="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName"></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('won\'t use the property name as a column heading if label is an empty string', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component :show-filter="false"
                    :data="[{ firstName: 'John' },{ id: 2, firstName: 'Paul' }]">
                    <table-column show="firstName" label=""></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('can accept a function to fetch the data', async () => {
    const serverResponse = () => {
      return {
        data: [{
          firstName: 'John'
        }, {
          id: 2,
          firstName: 'Paul'
        }],
      };
    };

    document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="${serverResponse}">
                    <table-column show="firstName" label="First name"></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    await Vue.nextTick();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('can add extra classes to the table, the cells and the headers', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John' },{ firstName: 'Paul' }]"
                    table-class="my-table"
                    thead-class="my-thead"
                    tbody-class="my-tbody"
                >
                    <table-column
                        show="firstName"
                        label="First name"
                        header-class="my-header"
                        cell-class="my-cell"
                    ></table-column>
                </table-component>
            </div>
        `;

    await createVm();

    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('can update columns', async () => {
    document.body.innerHTML = `
            <div id="app">
                <table-component :data="[{ firstName: 'John' },{ firstName: 'Paul' }]">
                    <table-column show="firstName" :label="label"></table-column>
                </table-component>
            </div>
        `;

    const vm = new Vue({
      el: '#app',
      data: {
        label: 'First name',
      },
    });

    await Vue.nextTick();

    expect(document.body.innerHTML).toMatchSnapshot();

    vm.label = 'Something else';

    await Vue.nextTick();

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

async function createVm() {
  Vue.component('app', Vue.extend(App));
  Vue.component('table-component', Vue.extend(TableComponent));
  Vue.component('table-column', Vue.extend(TableColumn));
  const vm = new Vue({
    el: '#app',
  });

  await Vue.nextTick(() => {});

  const table = vm.$children[0];

  return table;
}
