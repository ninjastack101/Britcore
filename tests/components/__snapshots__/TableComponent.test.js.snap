// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent can accept a function to fetch the data 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <div class="table-component__message">
    </div>
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent can add extra classes to the table, the cells and the headers 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table my-table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head my-thead">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort my-header"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body my-tbody">
          <tr>
            <td class="my-cell">
              John
            </td>
          </tr>
          <tr>
            <td class="my-cell">
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent can display nested properties 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent can mount 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent can update columns 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent can update columns 2`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              Something else
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent has an prop to disable the caption 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <!---->
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent has an prop to disable the filter 1`] = `

<div id="app">
  <div class="table-component">
    <!---->
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent supports a named slot to display a tfoot section 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <td>
              Name count:
            </td>
            <td>
              2
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent supports a scoped slot inside the table column 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder
             class="table-component__filter__field"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John slot
            </td>
          </tr>
          <tr>
            <td>
              Paul slot
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent will use the property name as a column heading if label is not set 1`] = `

<div id="app">
  <div class="table-component">
    <!---->
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              firstName
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;

exports[`TableComponent won't use the property name as a column heading if label is an empty string 1`] = `

<div id="app">
  <div class="table-component">
    <!---->
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table not sorted
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
  </div>
</div>

`;
